### ISSUE TITLE
Enhancement: [restrict-template-expressions] More permissive type check

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#8722]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/restrict-template-expressions/

### Description

Related: #6279

I have added the `allowArray` option to `restrict-template-expressions` in #8389.

Now, I'd like to discuss other things that may be useful to put in a string. The most common case for me is stringifying `URLSearchParams`, `Date` and `Buffer`s, but there could be potential for others as well.
Some questions for discussion:

- Do we add a new option `ignoredTypeNames`? (I have created a draft PR #8556 implementing this)
- What will be the default value for this option?
- What does this mean for the `no-base-to-string` rule, which is becoming increasingly similar to this rule?

### Fail

```TypeScript
declare const maybeParams: URLSearchParams | undefined;

fetch(`https://myapi.com?${maybeParams}`);
```

### Pass

```TypeScript
declare const definitelyParams: URLSearchParams;

fetch(`https://myapi.com?${definitelyParams}`);
```

**Comment by bradzacher** [comment]
Quick response:

> Do we add a new option ignoredTypeNames?

Nope. We do not want loose "type name a string" options.
https://github.com/typescript-eslint/typescript-eslint/discussions/6017

TL;DR - if you configure `ignoreTypeNames: ["Date"]` how does that differentiate between
```ts
declare const x: Date;
`${x}` // will result in '<..date time string here..>'
```
and
```ts
interface Date { prop: string }
declare const x: Date;
`${x}` // will result in '[object Object]'
```

**Comment by abrahamguo** [comment]
@bradzacher Does this mean that `ignoredTypeNames` in `no-base-to-string` is also on the chopping block to be removed, because of the same issue?

**Comment by abrahamguo** [comment]
What do you guys think about the comments on #6279?

(which, really, I don't think should be closed, since it hasn't really been fixed?)

**Comment by JoshuaKGoldberg** [comment]
I'm personally in favor of adding the standard format of _"allow specifically these types"_ to `restrict-template-expressions` (https://github.com/typescript-eslint/typescript-eslint/issues/8722#issuecomment-2007199042). Would like to see more input from @typescript-eslint/triage-team before making any decisions.

**Comment by JoshuaKGoldberg** [comment]
A 👍 from another team member and a past _"I'm ok with adding"_ from https://github.com/typescript-eslint/typescript-eslint/issues/6279#issuecomment-1457287495. Marking as accepting PRs!
